zcash_client_backend/src/fees/common.rs

```diff
@@ -358,10 +420,223 @@ where
+        );
+        let o_allowed = min(
+            o_dust.len(),
+            (o_outputs_len + o_change).saturating_sub(o_non_dust),
+        );
+
+        // We'll be spending the non-dust and allowed dust in each pool.
+        let t_req_inputs = t_non_dust + t_allowed;
+        let s_req_inputs = s_non_dust + s_allowed;
+        #[cfg(feature = "orchard")]
+        let o_req_inputs = o_non_dust + o_allowed;
+
+        // This calculates the hypothetical number of actions with given extra inputs,
+        // for ZIP 317 and the padding rules in effect. The padding rules for each
+        // pool are subtle (they also depend on `bundle_required` for example), so we
+        // must actually call them rather than try to predict their effect. To tell
+        // whether we can freely add an extra input from a given pool, we need to call
+        // them both with and without that input; if the number of actions does not
+        // increase, then the input is free to add.
+        let hypothetical_actions = |t_extra, s_extra, _o_extra| {
+            let s_spend_count = sapling
+                .bundle_type()
+                .num_spends(s_req_inputs + s_extra)
+                .map_err(ChangeError::BundleError)?;
+
+            let s_output_count = sapling
+                .bundle_type()
+                .num_outputs(s_req_inputs + s_extra, s_outputs_len + s_change)
+                .map_err(ChangeError::BundleError)?;
+
+            #[cfg(feature = "orchard")]
+            let o_action_count = orchard
+                .bundle_type()
+                .num_actions(o_req_inputs + _o_extra, o_outputs_len + o_change)
+                .map_err(ChangeError::BundleError)?;
+            #[cfg(not(feature = "orchard"))]
+            let o_action_count = 0;
+
+            // To calculate the number of unused actions, we assume that transparent inputs
+            // and outputs are P2PKH.
+            Ok(max(t_req_inputs + t_extra, t_outputs_len + t_change)
+                + max(s_spend_count, s_output_count)
+                + o_action_count)
+        };
+
+        // First calculate the baseline number of logical actions with only the definitely
+        // allowed inputs estimated above. If it is less than `grace_actions`, try to allocate
+        // a grace input first to transparent dust, then to Sapling dust, then to Orchard dust.
+        // If the number of actions increases, it was not possible to allocate that input for
+        // free. This approach is sufficient because at most one such input can be allocated,
+        // since `grace_actions` is at most 2 for ZIP 317 and there must be at least one
+        // logical action. (If `grace_actions` were greater than 2 then the code would still
+        // be correct, it would just not find all potential extra inputs.)
+
+        let baseline = hypothetical_actions(0, 0, 0)?;
+
+        let (t_extra, s_extra, o_extra) = if baseline >= grace_actions {
```

Comment:
Could this be refactored into something that doesn't involve such code complexity, or at least extracted into a function that can be unit tested?

Reply:
It's not really possible to make it much less complex, I think. If we want to unit-test `hypothetical_actions`, we can easily pull it out into a function then. None of this is public API surface, so we don't need to do that in the absence of an actual test.